File: heap/02_top_k_frequent_elements.py
"""
PROBLEM: Top K Frequent Elements (LeetCode 347)
DIFFICULTY: Medium
PATTERN: Heap - Min-Heap of size k by frequency

DESCRIPTION:
Given an integer array nums and an integer k, return the k most frequent elements.

EXAMPLE:
Input: nums = [1,1,1,2,2,3], k = 2 -> Output: [1, 2]
Input: nums = [1], k = 1 -> Output: [1]

TIME COMPLEXITY: O(n log k)
SPACE COMPLEXITY: O(n)
"""

import heapq
from typing import List
from collections import Counter


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = Counter(nums)
        # min-heap of (frequency, element); keep size k
        min_heap: List = []
        for num, cnt in freq.items():
            heapq.heappush(min_heap, (cnt, num))
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        return [num for _, num in min_heap]

    def topKFrequentBucket(self, nums: List[int], k: int) -> List[int]:
        freq = Counter(nums)
        buckets: List[List[int]] = [[] for _ in range(len(nums) + 1)]
        for num, cnt in freq.items():
            buckets[cnt].append(num)
        result: List[int] = []
        for i in range(len(buckets) - 1, -1, -1):
            result.extend(buckets[i])
            if len(result) >= k:
                return result[:k]
        return result[:k]


if __name__ == "__main__":
    sol = Solution()
    test_cases = [
        ([1, 1, 1, 2, 2, 3], 2),
        ([1], 1),
        ([4, 4, 4, 6, 6, 7], 2),
    ]
    for nums, k in test_cases:
        print(f"nums={nums}, k={k} -> {sol.topKFrequent(nums, k)}")

File: heap/02_top_k_frequent_elements.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <algorithm>
using namespace std;


/*
Problem Description:
Given an integer array nums and an integer k, return the k most frequent elements.

Sample Input/Output:
1) nums = [1,1,1,2,2,3], k = 2 -> Output: [1,2]
2) nums = [1], k = 1 -> Output: [1]
3) nums = [4,4,4,6,6,7], k = 2 -> Output: [4,6]

Code Explanation:
Count frequencies using a hash map. Keep a min-heap of size k on frequency.
When heap grows beyond k, remove the smallest frequency.
*/

vector<int> topKFrequent(vector<int> &nums, int k) {
    unordered_map<int, int> freq;
    for (int x : nums) freq[x]++;

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
    for (auto &entry : freq) {
        minHeap.push({entry.second, entry.first});
        if ((int)minHeap.size() > k) minHeap.pop();
    }

    vector<int> result;
    while (!minHeap.empty()) {
        result.push_back(minHeap.top().second);
        minHeap.pop();
    }
    reverse(result.begin(), result.end());
    return result;
}

void printVec(const vector<int> &v) {
    cout << "[";
    for (int i = 0; i < (int)v.size(); i++) {
        cout << v[i] << (i + 1 == (int)v.size() ? "" : ",");
    }
    cout << "]\n";
}

int main() {
    vector<int> a1 = {1,1,1,2,2,3};
    vector<int> a2 = {1};
    vector<int> a3 = {4,4,4,6,6,7};

    cout << "Output 1: "; printVec(topKFrequent(a1, 2));
    cout << "Output 2: "; printVec(topKFrequent(a2, 1));
    cout << "Output 3: "; printVec(topKFrequent(a3, 2));
    return 0;
}
